wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -0,0 +1,516 @@
+import com.squareup.wire.ProtoAdapter
+import com.squareup.wire.ProtoReader
+import com.squareup.wire.ProtoWriter
+import com.squareup.wire.WireEnum
+import com.squareup.wire.schema.EnumType
+import com.squareup.wire.schema.Field
+import com.squareup.wire.schema.MessageType
+import com.squareup.wire.schema.ProtoFile
+import com.squareup.wire.schema.ProtoType
+import com.squareup.wire.schema.Schema
+import com.squareup.wire.schema.Type
+import okio.ByteString
+import java.util.Locale
+
+class KotlinGenerator private constructor(
+  val schema: Schema,
+  private val nameToKotlinName: Map<ProtoType, ClassName>,
+  private val emitAndroid: Boolean
+) {
+  val nameAllocatorStore = mutableMapOf<Type, NameAllocator>()
+
+  /** Returns the Kotlin type for [protoType]. */
+  fun typeName(protoType: ProtoType) = nameToKotlinName.getValue(protoType)
+
+  /** Returns the full name of the class generated for `type`.  */
+  fun generatedTypeName(type: Type) = typeName(type.type())
+
+  fun generateType(type: Type): TypeSpec = when (type) {
+    is MessageType -> generateMessage(type)
+    is EnumType -> generateEnum(type)
+    else -> error("Unknown type $type")
+  }
+
+  private fun nameAllocator(message: Type): NameAllocator {
+    return nameAllocatorStore.getOrPut(message) {
+      NameAllocator().apply {
+        when (message) {
+          is EnumType -> {
+            newName("value", "value")
+            newName("ADAPTER", "ADAPTER")
+            message.constants().forEach { constant ->
+              newName(constant.name(), constant)
+            }
+          }
+          is MessageType -> {
+            newName("unknownFields", "unknownFields")
+            newName("ADAPTER", "ADAPTER")
+            newName("reader", "reader")
+            if (emitAndroid) {
+              newName("CREATOR", "CREATOR")
+            }
+            message.fields().forEach { field ->
+              newName(field.name(), field)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private fun generateMessage(type: MessageType): TypeSpec {
+    val className = typeName(type.type())
+
+    val classBuilder = TypeSpec.classBuilder(className)
+        .addModifiers(DATA)
+        .addType(generateAdapter(type))
+
+    if (emitAndroid) {
+      addAndroidCodeToMessage(classBuilder, type)
+    }
+
+    addMessageConstructor(type, classBuilder)
+
+    type.nestedTypes().forEach { classBuilder.addType(generateType(it)) }
+
+    return classBuilder.build()
+  }
+
+  /**
+   * Example
+   * ```
+   * data class Person(
+   *   val name: String,
+   *   val email: String? = null,
+   *   val phone: List<PhoneNumber> = emptyList(),
+   *   val unknownFields: ByteString = ByteString.EMPTY
+   * )
+   * ```
+   */
+  private fun addMessageConstructor(message: MessageType, classBuilder: TypeSpec.Builder) {
+    val constructorBuilder = FunSpec.constructorBuilder()
+    constructorBuilder.addAnnotation(AnnotationSpec.builder(JvmOverloads::class)
```

Comment:
nit: this isn’t in the example

Reply:
... and I’m not 100% sure we want it . .  .

Reply:
(the drawback is that adding any field is not backwards-compatible)

Reply:
Also, you can do

```
constructorBuilder.jvmOverloads()
```

Reply:
Good point!